components/movementsensor/gpsrtkserial/gpsrtkserial.go

```diff
@@ -371,47 +388,77 @@ func (g *rtkSerial) closePort() {
 	}
 }
 
-// receiveAndWriteSerial connects to NTRIP receiver and sends correction stream to the MovementSensor through serial.
-func (g *rtkSerial) receiveAndWriteSerial() {
-	defer g.activeBackgroundWorkers.Done()
+// ntripConnection connects to NTRIP stream.
+func (g *rtkSerial) ntripConnection() error {
 	if err := g.cancelCtx.Err(); err != nil {
-		return
+		return g.err.Get()
 	}
+	g.urlMutex.Lock()
 	err := g.connect(g.ntripClient.URL, g.ntripClient.Username, g.ntripClient.Password, g.ntripClient.MaxConnectAttempts)
 	if err != nil {
 		g.err.Set(err)
-		return
+		return g.err.Get()
 	}
 
 	if !g.ntripClient.Client.IsCasterAlive() {
 		g.logger.Infof("caster %s seems to be down", g.ntripClient.URL)
 	}
 
+	g.logger.Debug("gettting source table")
+	srcTable, err := g.ntripClient.Client.ParseSourcetable()
+	if err != nil {
+		g.logger.Errorf("failed to get source table: %v", err)
+	}
+	g.logger.Debug("got sourcetable, parsing it...")
+	g.isVirtualBase, err = findLineWithMountPoint(srcTable, g.ntripClient.MountPoint)
+	if err != nil {
+		g.logger.Errorf("can't find mountpoint in source table, found err %v\n", err)
+	}
+	g.urlMutex.Unlock()
 	err = g.openPort()
 	if err != nil {
 		g.err.Set(err)
-		return
+		return g.err.Get()
 	}
 
-	defer g.closePort()
+	g.w = bufio.NewWriter(g.correctionWriter)
+	g.logger.Debug("connecting to NTRIP stream........")
 
-	w := bufio.NewWriter(g.correctionWriter)
-	err = g.getStream(g.ntripClient.MountPoint, g.ntripClient.MaxConnectAttempts)
-	if err != nil {
-		g.err.Set(err)
-		return
+	if g.isVirtualBase {
+		g.rw, err = g.sendGGAMessage()
+		if err != nil {
+			g.err.Set(err)
+			return g.err.Get()
+		}
+		g.r = io.TeeReader(g.rw, g.w)
+	} else {
+		err = g.getStream(g.ntripClient.MountPoint, g.ntripClient.MaxConnectAttempts)
+		if err != nil {
+			g.err.Set(err)
+			return g.err.Get()
+		}
+
+		g.r = io.TeeReader(g.ntripClient.Stream, g.w)
 	}
 
-	r := io.TeeReader(g.ntripClient.Stream, w)
-	scanner := rtcm3.NewScanner(r)
+	return nil
+}
+
+// receiveAndWriteSerial connects to NTRIP receiver and sends correction stream to the MovementSensor through serial.
+func (g *rtkSerial) receiveAndWriteSerial() {
+	defer g.activeBackgroundWorkers.Done()
+	defer g.closePort()
```

Comment:
no - that is a larger refactor than we can do at this point, since it is reliant on the gpsnmea struct to open the port - which is something I want done eventually but not in this pr.